### PR TITLE
pre-commit: Add hooks for qmllint and qmlformat

### PR DESCRIPTION
**Pull request by Holzhaus**
This adds the [qmllint](https://doc-snapshots.qt.io/qt6-dev/qtquick-tools-and-utilities.html#qmllint) and [qmlformat](https://doc-snapshots.qt.io/qt6-dev/qtquick-tools-and-utilities.html#qmlformat) hooks to ensure our QML files are up to par.

Unfortunately, `qmlformat` is only part of Qt 5.15 and later. Therefore, we cannot expect our contributors to have it installed. Since it's a binary dependencies (i.e. not a python/npm package), we also can't provide convenient automatic installation.

Therefore, the qmlformat hook is restricted to the `manual` hook stage. This means that it's not run automatically when committing, only when you run:

    $ pre-commit run qmlformat --hook-stage manual --all-files

Not great and easy to forget, but we can let our pre-commit CI workflow assist us in noticing issues during PR review. However, this requires that our CI has Qt 5.15 available, which isn't the case with the current `ubuntu-latest` image.

Hence, I wrote a [small Dockerfile based on Arch Linux](https://github.com/Holzhaus/mixxx-ci-docker/tree/20210526) and [uploaded the image to DockerHub](https://hub.docker.com/layers/holzhaus/mixxx-ci/20210526/images/sha256-af57ff7576d943ce6ad0acabd2259c91730d13f2711fe226d91649e48456e920), so that we can use it in our pre-commit workflow.

TL;DR: With this PR, QML formatting issues  can be detected early on CI without requiring all contributors to install Qt 5.15 or later.

### CODE REVIEW

**Comment by poelzi** [comment]
What shall I say, that is why nix is superiour and does not require such hacks....

**Comment by Holzhaus** [comment]
> What shall I say, that is why nix is superiour and does not require such hacks....

We cannot force our contributors to use a particular OS (= one with a non-outdated Qt version), or set up nix or docker, or build Qt5 themselves locally.

I have no problems on my machine, as I use Qt 5.15 for quite some time from my distros official repos. The whole "hack" has only been added for people on outdated distros.

**Comment by uklotzde** [comment]
> What shall I say, that is why nix is superiour and does not require such hacks....

Please stay on topic. We don't need to fuel the *Linux distro wars*. Let's just work together to keep Mixxx platform-independent. Thank you.

**Comment by uklotzde** [comment]
Shall we add the corresponding Dockerfile to our repo, e.g. in `tools/ci`? The purpose of this Dockerfile and how it is supposed to be used could be added as a comment, everything self-contained.

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -8,18 +8,13 @@ jobs:
   pre-commit:
     name: Detecting code style issues
     runs-on: ubuntu-latest
+    container: holzhaus/mixxx-ci:20210526
```
Comment (uklotzde) [inline .github/workflows/pre-commit.yml:11]: We could then also add a comment here where the corresponding Dockerfile for generating this image could be found in the repo.

File: .github/workflows/pre-commit.yml [hunk 2]
```diff
@@ -28,16 +23,23 @@ jobs:
       # on Pull Requests!).
       env:
         SKIP: end-of-file-fixer,trailing-whitespace,clang-format,eslint,no-commit-to-branch
+      with:
+        # Enable extra hooks that require special system dependencies that we
+        # cannot expect to be installed on all contributor's systems.
```
Comment (uklotzde) [inline .github/workflows/pre-commit.yml:28]: Please add a note that the custom container image used for running this action contains those dependencies.

**Comment by Holzhaus** [comment]
> Now we have to maintain even more distribution specific stuff, regullarly build docker containers, push those etc, who takes care about that, are those containers CI built ?

We don't have to rebuild them unless we want to use a more recent Qt/clang-format version. 
The container shouldn't magically break or something.

It's fine that you're passionate about nix, but do you really suggest to required setting up nix in order to contribute to Mixxx? Because I don't see another way to get around this if we don't want this "hack".

**Comment by Holzhaus** [comment]
Comments added.

**Review by uklotzde — APPROVED** [review]
Thank you! LGTM

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -124,3 +124,23 @@ repos:
...
+  - id: qmlformat
+    name: qmlformat
+    entry: qmlformat -i
+    pass_filenames: true
+    require_serial: true
+    language: system
+    types: [text]
+    files: ^.*\.qml$
+    # Not enabled in commit stage by default, because qmlformat requires Qt
+    # 5.15 to be installed
```
[the remark this answers is not in the document]
Comment (Holzhaus) [reply .pre-commit-config.yaml:136]: You will notice when it runs on CI at least. I thought about that as well, but thr problem is that the hook will always pass in that case, and you don't really know it it actually passed or if the tool was just not found.

[the remark this answers is not in the document]
Comment (Holzhaus) [reply .pre-commit-config.yaml:136]: The alternative would be to not make any special workarounds. The hook will only run if you changed QML files anyway. If you really want to change QML files with qml format installed just skip the hook with `SKIP=qmlformat git commit`

That would also be an option. @poelzi @uklotzde what do you think?
